misc/git-hooks/pre-commit

```diff
@@ -15,7 +15,7 @@ set -e
 
 
 # Stash pending changes and revert them when script ends
-if [ $is_unclean -ne 0 ]; then
+if [ -z "${NO_STASH:-}" ] && [ $is_unclean -ne 0 ]; then
```

Comment:
I also like the introduction of the `NO_STASH` variable. I had it a couple times that I added changes via `git add -p` and wanted to commit them, but aborted the commit, and then got resulting "merge conflicts" when it was unstashing. 

Was are we stashing in the first place? for the format checks and similar?

Reply:
Any changes that are not `git add`-ed are stashed so that formatting and lints checks only the code that is included in the commit.